comp/core/workloadmeta/workloadmeta.go

```diff
@@ -0,0 +1,98 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package workloadmeta
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/comp/core/config"
+	"github.com/DataDog/datadog-agent/comp/core/log"
+	"github.com/DataDog/datadog-agent/pkg/util/common"
+	"go.uber.org/fx"
+)
+
+// store is a central storage of metadata about workloads. A workload is any
+// unit of work being done by a piece of software, like a process, a container,
+// a kubernetes pod, or a task in any cloud provider.
+type workloadmeta struct {
+	log    log.Component
+	config config.Component
+
+	// Store related
+	storeMut sync.RWMutex
+	store    map[Kind]map[string]*cachedEntity // store[entity.Kind][entity.ID] = &cachedEntity{}
+
+	subscribersMut sync.RWMutex
+	subscribers    []subscriber
+
+	collectorMut sync.RWMutex
+	candidates   map[string]Collector
+	collectors   map[string]Collector
+
+	eventCh chan []CollectorEvent
+
+	ongoingPullsMut sync.Mutex
+	ongoingPulls    map[string]time.Time // collector ID => time when last pull started
+}
+
+// InitHelper this should be provided as a helper to allow passing the component into
+// the inithook for additional start-time configutation.
+type InitHelper func(context.Context, Component) error
+
+type dependencies struct {
+	fx.In
+
+	Lc fx.Lifecycle
+
+	Log     log.Component
+	Config  config.Component
+	Catalog CollectorList `group:"workloadmeta"`
+
+	Params Params
+}
+
+func newWorkloadMeta(deps dependencies) Component {
+	candidates := make(map[string]Collector)
+	for _, c := range deps.Catalog {
+		if c.GetTargetCatalog() == deps.Params.AgentType {
```

Comment:
I think this pattern introduces a constraint where a collector must be in a single catalog. I don't think it's a blocker for this PR but perhaps we might want to relax this constraint